File: output/0695.岛屿的最大面积/0695-岛屿的最大面积.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        def dfs(i,j):
            if grid[i][j] != 1:
                return 0
            grid[i][j] = 0
            a = b = c = d = 0
            if i>0:
                a = dfs(i-1,j)
            if j>0:
                b = dfs(i,j-1)
            if i<m-1:
                c = dfs(i+1,j)
            if j < n-1:
                d = dfs(i,j+1)
            return a+b+c+d+1
        res = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j]:
                    res  = max(res,dfs(i,j))
        return res

File: 答题/全部/0695.岛屿的最大面积/0695-岛屿的最大面积.go
func maxAreaOfIsland(grid [][]int) int {
    res := 0
    var dfs func(i, j int) int
    dfs = func(i, j int) int {
        if i < 0 || j < 0 || i == len(grid) || j == len(grid[0]) || grid[i][j] == 0 {
            return 0
        }
        grid[i][j] = 0
        sum := 1
        sum += dfs(i-1, j)
        sum += dfs(i+1, j)
        sum += dfs(i, j-1)
        sum += dfs(i, j+1)
        return sum
    }
    for i := 0; i < len(grid); i++ {
        for j := 0; j < len(grid[0]); j++ {
            if grid[i][j] == 1 {
                res = max(res, dfs(i, j))
            }
        }
    }


    return res
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}